### ISSUE TITLE
Plumb page width from analysis_options.yaml to formatter when formatting in analysis_server

### ISSUE THREAD

**Issue by munificent** [dart-lang/sdk#56864]
My understanding is that analysis_server uses the formatter directly as a library so that it can format code as it exists in the IDE and not on disk.

I recently added support to the `dart format` CLI (but not the dart_style library API) to look for a surrounding `analysis_options.yaml` file and configure the page width from there if present. The schema looks like:

```dart
formatter:
  page_width: 123
```

To be consistent, analysis_server should also plumb that same page width in when it formats a file in memory. The library API already supports a `pageWidth` optional parameter, so this can be implemented whenever.

Let me know if you have any questions and I can fill you in on all the details. If you want a reference, the `dart format` implementation is [here](https://github.com/dart-lang/dart_style/tree/main/lib/src/analysis_options) with [these tests](https://github.com/dart-lang/dart_style/tree/main/test/analysis_options). (I harvested this code from analyzer initially, but ended up rewriting a lot because the YamlNode API is... weird.)

Thank you!

cc @DanTup

**Comment by DanTup** [comment]
@bwilkerson @munificent both the LSP and legacy server have ways to configure page width already - which should take priority when both are provided?

My feeling is that `analysis_options.yaml` should be priority, so if in VS Code you set your own default to 120 but clone a project using 100, you will get 100. However, in the legacy protocol the value is passed to the individual `edit.format` call and so ignoring that value (because there's one in analysis_options) seems slightly weirder.

I presume the rules for overriding `analysis_options.yaml` in `dart format` are the same as what's been implemented for the server - eg. an `analysis_options.yaml` completely prevents use of values from any ancestor unless it was `include`d? I think [these tests](https://github.com/dart-lang/dart_style/blob/618883f734f8af5df2a0e7ae47dec970fff016f8/test/analysis_options/analysis_options_file_test.dart#L41) confirm that, although I couldn't see a test that uses `include` and specifically checks the `pageWidth`.

**Comment by DanTup** [comment]
> The only reason it's a request parameter is because we didn't provide any kind of configuration support. I don't have any concerns about treating the user preference the same across the IDEs, and I think it might be confusing if they were treated differently.

Thanks - I'll do this. I think your other answers align with the assumptions I made the first CL at https://dart-review.googlesource.com/c/sdk/+/388822, though I am just accepting >= 1 because @munificent didn't mention a minimum - but I can update the validation if there is (if there is, perhaps it should be exposed by the formatter as a constant to avoid us duplicating it).

**Comment by DanTup** [comment]
This was done via 5d84439f5d09eaeb2c544dc60247077f68f0b995 and 75d15bbfa36411b869b10ab161c60bc911e3437a.

**Comment by munificent** [comment]
> My feeling is that `analysis_options.yaml` should be priority, so if in VS Code you set your own default to 120 but clone a project using 100, you will get 100. 

I agree. It's not good to be in a situation where your local preferences override a committed configuration because then any changes you make are likely to break CI or cause large diffs when you upload a PR.

> However, in the legacy protocol the value is passed to the individual `edit.format` call and so ignoring that value (because there's one in analysis_options) seems slightly weirder.

I don't know what this means, but that's OK. :)

> I presume the rules for overriding `analysis_options.yaml` in `dart format` are the same as what's been implemented for the server - eg. an `analysis_options.yaml` completely prevents use of values from any ancestor unless it was `include`d?

Yes. I borrowed the implementation and semantics from server. A nearer `analysis_options.yaml` file always completely shadows an outer one, even if it doesn't have a particular setting.

> I think [these tests](https://github.com/dart-lang/dart_style/blob/618883f734f8af5df2a0e7ae47dec970fff016f8/test/analysis_options/analysis_options_file_test.dart#L41) confirm that

Right.

> although I couldn't see a test that uses `include` and specifically checks the `pageWidth`.

[These tests](https://github.com/dart-lang/dart_style/blob/main/test/cli/page_width_test.dart) are more like integration tests and validate that the page width makes its way from the `analysis_options.yaml` file all the way through to formatting.

> 1. is this valid?
> 
> ```yaml
> formatter:
>   page_width:
> ```

In the formatter's code for extracting a page_width from `analysis_options.yaml`, it silently fails if at any point it can't get to an integer for the `page_width` inside `formatter`. So in this case, I think YAML gives you `null` for the value? It should then silently reject that.

> 2. Is 0 valid?

It's definitely not going to do anything useful, but the formatter will let you try. :)

Thank you for the quick turnaround!

**Comment by DanTup** [comment]
> 2. Is 0 valid?
> It's definitely not going to do anything useful, but the formatter will let you try. :)

I treated 0 as invalid, so if the user types this they will get a diagnostic (same as if they type a negative value), and we will ignore it (that is, we'll pass nothing to the formatter, or the globally configured value from the IDE).

I think this is fine, but if `dart format` would pass the zero to the formatter and you think we should do the same here for consistency, let me know. (I guess the same goes for negatives - are you treating them as entirely invalid and ignoring them and using 80, or would it do the same as `0/1`?)

**Comment by munificent** [comment]
> I think this is fine, but if `dart format` would pass the zero to the formatter and you think we should do the same here for consistency, let me know.

I honestly don't think it matters much one way or the other. For what it's worth, `dart format` will let you pass zero or a negative number and will then try to format at that (nonsensical width). I suspect that what you'll get is just maximally line split output.
